cmd/util/ledger/migrations/cadence_value_validation.go

```diff
@@ -0,0 +1,496 @@
+		}
+
+		newValue := newStorageMap.ReadValue(nopMemoryGauge, interpreter.StringStorageMapKey(stringKey))
+
+		if !cadenceValueEqual(oldRuntime.Interpreter, oldValue, newRuntime.Interpreter, newValue) {
+			return fmt.Errorf("failed to validate domain %s, key %s: old value %v (%T), new value %v (%T)", domain, key, oldValue, oldValue, newValue, newValue)
+		}
+	}
+
+	return nil
+}
+
+func cadenceValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v interpreter.Value,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) bool {
+	switch v := v.(type) {
+	case *interpreter.ArrayValue:
+		return cadenceArrayValueEqual(vInterpreter, v, otherInterpreter, other)
+
+	case *interpreter.CompositeValue:
+		return cadenceCompositeValueEqual(vInterpreter, v, otherInterpreter, other)
+
+	case *interpreter.DictionaryValue:
+		return cadenceDictionaryValueEqual(vInterpreter, v, otherInterpreter, other)
+
+	case *interpreter.SomeValue:
+		return cadenceSomeValueEqual(vInterpreter, v, otherInterpreter, other)
+
+	default:
+		oldValue, ok := v.(interpreter.EquatableValue)
+		if !ok {
+			return false
+		}
+		return oldValue.Equal(nil, interpreter.EmptyLocationRange, other)
+	}
+}
+
+func cadenceSomeValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v *interpreter.SomeValue,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) bool {
+	otherSome, ok := other.(*interpreter.SomeValue)
+	if !ok {
+		return false
+	}
+
+	innerValue := v.InnerValue(vInterpreter, interpreter.EmptyLocationRange)
+
+	otherInnerValue := otherSome.InnerValue(otherInterpreter, interpreter.EmptyLocationRange)
+
+	return cadenceValueEqual(vInterpreter, innerValue, otherInterpreter, otherInnerValue)
+}
+
+func cadenceArrayValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v *interpreter.ArrayValue,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) bool {
+	otherArray, ok := other.(*interpreter.ArrayValue)
+	if !ok {
+		return false
+	}
+
+	count := v.Count()
+	if count != otherArray.Count() {
+		return false
+	}
+
+	if v.Type == nil {
+		if otherArray.Type != nil {
+			return false
+		}
+	} else if otherArray.Type == nil ||
+		!v.Type.Equal(otherArray.Type) {
+		return false
+	}
+
+	for i := 0; i < count; i++ {
+		element := v.Get(vInterpreter, interpreter.EmptyLocationRange, i)
+		otherElement := otherArray.Get(otherInterpreter, interpreter.EmptyLocationRange, i)
+
+		if !cadenceValueEqual(vInterpreter, element, otherInterpreter, otherElement) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func cadenceCompositeValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v *interpreter.CompositeValue,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) bool {
+	otherComposite, ok := other.(*interpreter.CompositeValue)
+	if !ok {
+		return false
+	}
+
+	if !v.StaticType(vInterpreter).Equal(otherComposite.StaticType(otherInterpreter)) ||
```

Comment:
Just to confirm, passing this check means `v` and `other` have the same fields, right? In other words, no need to check if `other` has additional fields that `v` doesn't have?

Reply:
Great catch!  :+1: 

`StaticType` doesn't check fields, so I added a temporary workaround for this.

I'll open PR to improve this workaround when [Cadence PR 3002](https://github.com/onflow/cadence/pull/3002) is merged and the function I requested becomes available.